Add start_time and last_updated fields to _scheduler/ endpoints output

nickva opened the pull request:
Also for consistency add _replication_start_time field to completed and failed
replication document updates.

`start_time` is defined as the time since the replication first was noticed by
the replicator. For a document-based replication it is when the document update
event reached couch replicator and the replication record was parsed from it.
For a _replicate replication it is the time when the POST request body was
parsed.

`last_updated` is time of last state update.

sagelywizard commented:
Do we really need a `last_updated` field, given that the last entry in `history` has a timestamp?

nickva commented:
@sagelywizard oh you mean in the json output of jobs? I added it mainly for consistency. But you're right, we probably don't need it. I'll remove it

nickva committed:
Add start_time and last_updated fields to _scheduler/ endpoints output

Also for consistency add _replication_start_time field to completed and failed
replication document updates.

`start_time` is defined as the time since the replication first was noticed by
the replicator. For a document-based replication it is when the document update
event reached couch replicator and the replication record was parsed from it.
For a _replicate replication it is the time when the POST request body was
parsed.

`last_updated` is time of last state update.

sagelywizard commented on src/couch_replicator.erl:
```diff
@@ -269,11 +278,17 @@ doc_from_db(RepDb, DocId, UserCtx) ->
+                    StateT = couch_util:get_value(<<"_replication_state_time">>, Props, null),
+                    {StartT, StateT, Info, 1};
                 _OtherState ->
                     {null, 0}
```
This is a badmatch.

nickva replied:
oops, sorry, you're right! fixed

sagelywizard commented on src/couch_replicator.erl:
```diff
@@ -269,11 +278,17 @@ doc_from_db(RepDb, DocId, UserCtx) ->
-                    {couch_util:get_value(<<"_replication_state_reason">>, Props, null), 1};
+                    Info = couch_util:get_value(<<"_replication_state_reason">>, Props, null),
+                    StartT = couch_util:get_value(<<"_replication_start_time">>, Props, null),
+                    StateT = couch_util:get_value(<<"_replication_state_time">>, Props, null),
```
Can we pull these `StartT` and `StateT` `get_value`s out of this case?

nickva replied:
good idea

sagelywizard commented on src/couch_replicator.erl:
```diff
@@ -204,10 +205,16 @@ stream_active_docs_info([Node | Nodes], Cb, UserAcc, States) ->
 handle_replicator_doc_query({row, Props}, {Db, Cb, UserAcc, States}) ->
-    DocId = couch_util:get_value(id, Props),
-    DocStateBin = couch_util:get_value(key, Props),
+    DocId = get_value(id, Props),
```
It's my preference that these are left as-is. It's yucky that we're using an import in the first place and I don't see much point in changing it.

nickva replied:
I'll change it back

sagelywizard commented on src/couch_replicator.erl:
```diff
@@ -204,10 +205,16 @@ stream_active_docs_info([Node | Nodes], Cb, UserAcc, States) ->
+    {StartTime, StateTime, StateInfo} = case MapValue of
+        [StartT, StateT, Info] ->
+            {StartT, StateT, Info};
+        _Other ->
```
Why're you adding this extra case clause? IIUC the view can only emit a 3-element list.

nickva replied:
The view used to emit something else and we already deployed on bigblue was wondering if  it is better to handle previous view results there instead of crashing?

sagelywizard replied:
We should probably either add a comment or an explicit "upgrade" clause.

nickva replied:
Good idea. Will add a comment

sagelywizard commented on src/couch_replicator.erl:
```diff
@@ -207,7 +208,15 @@ handle_replicator_doc_query({row, Props}, {Db, Cb, UserAcc, States}) ->
     DocStateBin = couch_util:get_value(key, Props),
     DocState = erlang:binary_to_existing_atom(DocStateBin, utf8),
-    StateInfo = couch_util:get_value(value, Props),
+    MapValue = get_value(value, Props),
```
Can we change this back to use the module name as well?

nickva replied:
Will do

sagelywizard commented on src/couch_replicator_scheduler.erl:
```diff
@@ -57,6 +57,7 @@
           monitor :: undefined | reference() | '_',
           history :: history() | '_'}).
 
+
```
Could you remove this bit of whitespace addition?

nickva replied:
Oops, sorry. Updated

sagelywizard commented on src/couch_replicator_doc_processor.erl:
```diff
@@ -239,7 +242,8 @@ worker_returned(Ref, Id, {temporary_error, Reason}) ->
             errcnt = ErrCnt + 1,
-            worker = nil
+            worker = nil,
+            last_updated = os:timestamp()
```
Hrm. It's a little ugly that this is set in `worker_returned`, but it's set in `update_docs_row` for the other `worker_returned` clause. It'd be nice if they were set in the same spot, or at least at the same function call depth. Thoughts?

nickva replied:
Ah good point. Let's pull that up into worker_return then. All cases of update_docs_row update it so it can be done in one place only

sagelywizard reviewed (approved):
🏇
